basket/news/views.py

```diff
@@ -119,6 +123,85 @@ def ratelimited(request, e):
     }, 429)
 
 
+def generate_fxa_state():
+    return os.urandom(16).encode('hex')
+
+
+@require_safe
+def fxa_start(request):
+    fxa_oauth = get_fxa_clients()[0]
+    fxa_state = request.session['fxa_state'] = generate_fxa_state()
+    redirect_uri = request.build_absolute_uri('/fxa/success/')
+    redirect_to = fxa_oauth.get_redirect_url(state=fxa_state,
+                                             redirect_uri=redirect_uri,
+                                             scope='profile',
+                                             email=request.GET.get('email'))
+    return HttpResponseRedirect(redirect_to)
+
+
+@require_safe
+def fxa_success(request):
+    # remove state from session to prevent multiple attempts
+    sess_state = request.session.pop('fxa_state', None)
+    if sess_state is None:
+        # need a better error state here
+        return HttpResponseJSON({
+            'status': 'error',
+            'message': 'no state found',
+        }, 400)
+
+    code = request.GET.get('code')
+    state = request.GET.get('state')
+    if not (code and state):
+        # need a better error state here
+        return HttpResponseJSON({
+            'status': 'error',
+            'message': 'no code or state returned',
+        }, 400)
+
+    if sess_state != state:
+        # need a better error state here
+        return HttpResponseJSON({
+            'status': 'error',
+            'message': 'states do not match',
+        }, 400)
+
+    fxa_oauth, fxa_profile = get_fxa_clients()
+    try:
+        access_token = fxa_oauth.trade_code(code)['access_token']
+        user_profile = fxa_profile.get_profile(access_token)
+    except Exception:
+        sentry_client.captureException()
+        # need a better error state here
+        return HttpResponseJSON({
+            'status': 'error',
+            'message': 'error communicating with FxA',
+        }, 500)
+
+    # TODO use user_profile['uid'] to check account in SFMC
```

Comment:
Just verifying - this TODO is to check for an existing record/token in SFMC, and if found, use that token for the `get_user_data` call below?

Reply:
This is a note for a nice-to-have thing. In SFMC we have a table with email and FxA_ID. This is updated via messages we receive from FxA on a queue when people verify their email on a new account, but it's not the source of record and no sync process is perfect. So this is an opportunity to verify and correct our record of these users. I don't think we have to do that, but we could and maybe should. It's just not important for v1 of this and time is short.